File: src/test/java/es/leanmind/csv/CsvFilterShould.java
package es.leanmind.csv;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

class CsvFilterShould {

    public final String headerLine = "Num_factura, Fecha, Bruto, Neto, IVA, IGIC, Concepto, CIF_cliente, NIF_cliente";
    public final String emptyField = "";
    public final List<String> noInvoiceLines = List.of(headerLine);

    @Test
    void allow_for_correct_lines_only() {
        var ivaTax = "19";
        var invoiceLines = fileWithOneInvoiceLineHaving(ivaTax, emptyField);
        var incorrectLine = "A,line,missing,fields,and,wrong,formats,";
        invoiceLines.add(incorrectLine);

        var correctLines = Csv.filter(invoiceLines);

        assertThat(correctLines).isEqualTo(
                fileWithOneInvoiceLineHaving(ivaTax, emptyField)
        );
    }

    @Test
    void exclude_lines_with_both_tax_fields_populated() {
        var ivaTax = "19";
        var igicTax = "8";
        var invoiceLines = fileWithOneInvoiceLineHaving(ivaTax, igicTax);

        var correctLines = Csv.filter(invoiceLines);

        assertThat(correctLines).isEqualTo(noInvoiceLines);
    }

    @Test
    void at_least_one_tax_is_required() {
        var invoiceLines = fileWithOneInvoiceLineHaving(emptyField, emptyField);

        var correctLines = Csv.filter(invoiceLines);

        assertThat(correctLines).isEqualTo(noInvoiceLines);
    }

    @Test
    void exclude_lines_with_non_decimal_tax_fields() {
        var wrongTaxType = "XYZ";
        var invoiceLines = fileWithOneInvoiceLineHaving(wrongTaxType, emptyField);

        var correctLines = Csv.filter(invoiceLines);

        assertThat(correctLines).isEqualTo(noInvoiceLines);
    }

    @Test
    void test_exclude_lines_with_both_taxes_even_with_wrong_format() {
        var wrongIvaType = "XYZ";
        var igicTax = "8";
        var invoiceLines = fileWithOneInvoiceLineHaving(wrongIvaType, igicTax);

        var correctLines = Csv.filter(invoiceLines);

        assertThat(correctLines).isEqualTo(noInvoiceLines);
    }

    private ArrayList<String> fileWithOneInvoiceLineHaving(String ivaTax, String igicTax) {
        var invoiceId = "1";
        var invoiceDate = "02/05/2019";
        var grossAmount = "1000";
        var netAmount = "810";
        var concept = "ACER Laptop";
        var cif = "B76430134";
        var nif = "";
        String invoiceLine =
                String.join(",",
                        invoiceId,
                        invoiceDate,
                        grossAmount,
                        netAmount,
                        ivaTax,
                        igicTax,
                        concept,
                        cif,
                        nif
                );
        return new ArrayList<>(List.of(headerLine, invoiceLine));
    }
}

File: tdd-kotlin/kotlin/src/test/kotlin/examples/CsvFilterShould.kt
package examples

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class CsvFilterShould {
    private val headerLine = "Num_factura, Fecha, Bruto, Neto, IVA, IGIC, Concepto, CIF_cliente, NIF_cliente"
    private lateinit var filter: CsvFilter
    private val emptyDataFile = listOf(headerLine)
    private val emptyField = ""

    @BeforeEach
    fun setup() {
        filter = CsvFilter()
    }

    @Test
    fun allow_for_correct_lines_only() {
        val lines = fileWithOneInvoiceLineHaving(concept = "a correct line with irrelevant data")
        val result = filter.apply(lines)

        assertThat(result).isEqualTo(lines)
    }

    @Test
    fun exclude_lines_with_both_tax_fields_populated_as_they_are_exclusive() {
        val result = filter.apply(
            fileWithOneInvoiceLineHaving(ivaTax = "19", igicTax = "8")
        )

        assertThat(result).isEqualTo(emptyDataFile)
    }

    @Test
    fun exclude_lines_with_both_tax_fields_empty_as_one_is_required() {
        val result = filter.apply(
            fileWithOneInvoiceLineHaving(ivaTax = emptyField, igicTax = emptyField)
        )

        assertThat(result).isEqualTo(emptyDataFile)
    }

    @Test
    fun exclude_lines_with_non_decimal_tax_fields() {
        val result = filter.apply(
            fileWithOneInvoiceLineHaving(ivaTax = "XYZ")
        )

        assertThat(result).isEqualTo(emptyDataFile)
    }

    @Test
    fun exclude_lines_with_both_tax_fields_populated_even_if_non_decimal() {
        val result = filter.apply(
            fileWithOneInvoiceLineHaving(ivaTax = "XYZ", igicTax = "12")
        )

        assertThat(result).isEqualTo(emptyDataFile)
    }

    private fun fileWithOneInvoiceLineHaving(
        ivaTax: String = "19",
        igicTax: String = emptyField,
        concept: String = "irrelevant"
    ): List<String> {
        val invoiceId = "1"
        val invoiceDate = "02/05/2019"
        val grossAmount = "1000"
        val netAmount = "810"
        val cif = "B76430134"
        val nif = emptyField
        val formattedLine = listOf(
            invoiceId,
            invoiceDate,
            grossAmount,
            netAmount,
            ivaTax,
            igicTax,
            concept,
            cif,
            nif
        ).joinToString(",")
        return listOf(headerLine, formattedLine)
    }
}
